File: SkyEditor.Core.Windows/IOProvider.vb
﻿Imports System.IO
Imports Microsoft.VisualBasic.Devices

Public Class IOProvider
    Inherits SkyEditor.Core.IO.IOProvider

    Public Overrides Sub CopyFile(sourceFilename As String, destinationFilename As String)
        File.Copy(sourceFilename, destinationFilename, True)
    End Sub

    Public Overrides Sub DeleteFile(filename As String)
        File.Delete(filename)
    End Sub

    Public Overrides Sub WriteAllBytes(filename As String, data() As Byte)
        File.WriteAllBytes(filename, data)
    End Sub

    Public Overrides Sub WriteAllText(filename As String, data As String)
        File.WriteAllText(filename, data)
    End Sub

    Public Overrides Function CanLoadFileInMemory(fileSize As Long) As Boolean
        Return (New ComputerInfo).AvailablePhysicalMemory > (fileSize + 500 * 1024 * 1024)
    End Function

    Public Overrides Function DirectoryExists(directoryPath As String) As Boolean
        Return Directory.Exists(directoryPath)
    End Function

    Public Overrides Function fileExists(Filename As String) As Boolean
        Return File.Exists(Filename)
    End Function

    Public Overrides Function GetFileLength(filename As String) As Long
        Return (New FileInfo(filename)).Length
    End Function

    Public Overrides Function GetTempFilename() As String
        Return Path.GetTempFileName
    End Function

    Public Overrides Function OpenFile(filename As String) As Stream
        Return File.Open(filename, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Read)
    End Function

    Public Overrides Function OpenFileReadOnly(filename As String) As Stream
        Return File.Open(filename, FileMode.OpenOrCreate, FileAccess.Read, FileShare.ReadWrite)
    End Function

    Public Overrides Function OpenFileWriteOnly(filename As String) As Stream
        Return File.Open(filename, FileMode.OpenOrCreate, FileAccess.Write)
    End Function

    Public Overrides Function ReadAllBytes(filename As String) As Byte()
        Return File.ReadAllBytes(filename)
    End Function

    Public Overrides Function ReadAllText(filename As String) As String
        Return File.ReadAllText(filename)
    End Function
End Class

File: SkyEditor.Core.Android/IOProvider.cs
using System;
using System.IO;

namespace SkyEditor.Core.Android
{
    public class IOProvider : SkyEditor.Core.IO.IOProvider
    {
        public override bool CanLoadFileInMemory(long fileSize)
        {
            //Todo: Figure out how to determine how much RAM is available.
            //In the mean time, we'll only allow in-memory file caching if it's 5 MB or less.
            return fileSize <= (5 * 1024 * 1024);
        }

        public override void CopyFile(string sourceFilename, string destinationFilename)
        {
            File.Copy(sourceFilename, destinationFilename, true);
        }

        public override void CreateDirectory(string path)
        {
            Directory.CreateDirectory(path);
        }

        public override void DeleteDirectory(string path)
        {
            Directory.Delete(path, true);
        }

        public override void DeleteFile(string filename)
        {
            File.Delete(filename);
        }

        public override bool DirectoryExists(string directoryPath)
        {
            return Directory.Exists(directoryPath);
        }

        public override bool FileExists(string filename)
        {
            return File.Exists(filename);
        }

        public override string[] GetDirectories(string path, bool topDirectoryOnly)
        {
            if (topDirectoryOnly)
            {
                return Directory.GetDirectories(path, "*", SearchOption.TopDirectoryOnly);
            }
            else
            {
                return Directory.GetDirectories(path, "*", SearchOption.AllDirectories);
            }
        }

        public override long GetFileLength(string filename)
        {
            return (new FileInfo(filename).Length);
        }

        public override string[] GetFiles(string path, string searchPattern, bool topDirectoryOnly)
        {
            if (topDirectoryOnly)
            {
                return Directory.GetFiles(path, searchPattern, SearchOption.TopDirectoryOnly);
            }
            else
            {
                return Directory.GetFiles(path, searchPattern, SearchOption.AllDirectories);
            }
        }

        public override string GetTempDirectory()
        {
            var tempDir = Path.Combine(Path.GetTempPath(), "SkyEditor", Guid.NewGuid().ToString());
            if (!DirectoryExists(tempDir))
            {
                CreateDirectory(tempDir);
            }
            return tempDir;
        }

        public override string GetTempFilename()
        {
            return Path.GetTempFileName();
        }

        public override Stream OpenFile(string filename)
        {
            return File.Open(filename, FileMode.OpenOrCreate, FileAccess.ReadWrite, FileShare.Read);
        }

        public override Stream OpenFileReadOnly(string filename)
        {
            return File.Open(filename, FileMode.OpenOrCreate, FileAccess.Read, FileShare.ReadWrite);
        }

        public override Stream OpenFileWriteOnly(string filename)
        {
            return File.Open(filename, FileMode.OpenOrCreate, FileAccess.Write);
        }

        public override byte[] ReadAllBytes(string filename)
        {
            return File.ReadAllBytes(filename);
        }

        public override string ReadAllText(string filename)
        {
            return File.ReadAllText(filename);
        }

        public override void WriteAllBytes(string filename, byte[] data)
        {
            File.WriteAllBytes(filename, data);
        }

        public override void WriteAllText(string filename, string data)
        {
            File.WriteAllText(filename, data);
        }
    }
}
